File: utilities/nonvisualstudio/elixir/Implementation/Math.ex
# To use:
# Code.require_file "../../../../Utilities/nonvisualstudio/elixir/Implementation/Math.ex", __DIR__

defmodule Math do
  use Bitwise

  def gcd(left, right) do
    leftRemRight = rem(left, right)
    cond do
      leftRemRight == 0 ->
        right
      true ->
        gcd(right, leftRemRight)
    end
  end

  def lcm(left, right) do
    div((left * right), gcd(left,right))
  end

  def pow(n, exponent) do
    Stream.cycle([n])
    |> Enum.take(exponent)
    |> Enum.reduce(fn(elem, acc) -> elem * acc end)
  end

  #http://en.wikipedia.org/wiki/Modular_exponentiation
  def powMod(base, exponent, modulus) do
    _powMod(rem(base, modulus), exponent, modulus, 1)
  end

  defp _powMod(_, 0, _, result), do: result
  defp _powMod(base, exponent, modulus, result) do
    nextResult = _nextResult(base, exponent, modulus, result)
    nextExponent = exponent >>> 1
    nextBase = rem(base * base, modulus)

    _powMod(nextBase, nextExponent, modulus, nextResult)
  end

  defp _nextResult(base, exponent, modulus, currentResult) when rem(exponent, 2) === 1 do
    rem(currentResult * base, modulus)
  end
  defp _nextResult(_, _, _, currentResult), do: currentResult
end

File: utilities/nonvisualstudio/haskell/Implementation/Math.hs
-- to use - in ghci call use the following path
-- c:\Users\Thomas\Documents\GitHub\practice\utilities\nonvisualstudio\haskell\Implementation\Math.hs
-- import Math

module Math where

  import Data.Bits

  powMod :: Integer -> Integer -> Integer -> Integer
  powMod base theExponent modulus = powModRecur (base `mod` modulus) theExponent 1
    where
      powModRecur _ 0 result = result
      powModRecur currentBase currentExp currentResult = powModRecur nextBase nextExp nextResult
        where
          nextResult | currentExp `mod` 2 == 1 = (currentResult * currentBase) `mod` modulus
                     | otherwise               = currentResult
          nextExp   = shiftR currentExp 1
          nextBase  = (currentBase * currentBase) `mod` modulus
